File: src/reachability.rs
//! 状态类可达图
//!
//! BFS 探索构建 P-PTPN 状态类可达图

use crate::dbm::{Dbm, INF_TIME};
use crate::ptpn::{MatrixPTPN, INF};
use crate::state_class::{StateClass, TransitionEdge};
use petgraph::stable_graph::StableDiGraph;
use petgraph::visit::{EdgeRef, IntoEdgeReferences};
use std::collections::{HashMap, HashSet, VecDeque};
use std::fs;
use std::path::Path;
use tracing::{debug, info};

/// 状态类唯一键 (用于去重)
#[derive(Hash, Eq, PartialEq, Clone)]
struct StateKey {
    marking: Vec<i32>,
    z1_repr: Vec<Vec<i32>>,
    z2_repr: Vec<Vec<i32>>,
}

impl StateKey {
    fn from(state: &StateClass) -> Self {
        let z1_repr = (0..state.z1.size())
            .map(|i| {
                (0..state.z1.size())
                    .map(|j| state.z1.get_constraint(i, j))
                    .collect()
            })
            .collect();
        let z2_repr = (0..state.z2.size())
            .map(|i| {
                (0..state.z2.size())
                    .map(|j| state.z2.get_constraint(i, j))
                    .collect()
            })
            .collect();
        Self {
            marking: state.marking.clone(),
            z1_repr,
            z2_repr,
        }
    }
}

/// 可达图统计
#[derive(Debug, Default, Clone)]
pub struct ReachabilityStats {
    pub total_states: usize,
    pub total_transitions: usize,
    pub enabled_transitions_count: usize,
    pub pruned_states_count: usize,
}

/// 状态类可达图
pub struct StateClassReachabilityGraph {
    ptpn: MatrixPTPN,
    graph: StableDiGraph<StateClass, TransitionEdge>,
    initial_vertex: Option<petgraph::graph::NodeIndex>,
    stats: ReachabilityStats,
    state_to_vertex: HashMap<StateKey, petgraph::graph::NodeIndex>,
    next_state_id: usize,
    pruning_enabled: bool,
}

impl StateClassReachabilityGraph {
    pub fn new(ptpn: MatrixPTPN) -> Self {
        let graph = StableDiGraph::new();
        Self {
            ptpn,
            graph,
            initial_vertex: None,
            stats: ReachabilityStats::default(),
            state_to_vertex: HashMap::new(),
            next_state_id: 0,
            pruning_enabled: false,
        }
    }

    pub fn set_pruning_enabled(&mut self, enabled: bool) {
        self.pruning_enabled = enabled;
    }

    pub fn build(&mut self, max_states: usize) -> usize {
        self.stats = ReachabilityStats::default();
        self.state_to_vertex.clear();

        let s0 = self.create_initial_state_class();
        let canonical_s0 = self.canonicalize(&s0);

        let s0_vertex = self.find_or_add_vertex(&s0);
        self.initial_vertex = Some(s0_vertex);
        self.stats.total_states = 1;

        let key = StateKey::from(&s0);
        self.state_to_vertex.insert(key, s0_vertex);

        let mut queue = VecDeque::new();
        queue.push_back(canonical_s0);

        let mut iteration = 0;
        while let Some(cur) = queue.pop_front() {
            iteration += 1;
            let u = self.find_or_add_vertex(&cur);

            debug!("[状态 {}] 处理", cur.state_id);

            let chosen = self.select_per_core(&cur.enabled);
            self.stats.enabled_transitions_count += chosen.len();

            let mut scheduled = cur.copy();
            self.apply_preemption(&chosen, &mut scheduled);

            let mut dt = 0.0;
            if self.maximal_time_elapse(&mut scheduled, &mut dt) {
                debug!("  执行最大化时间推进: dt = {}", dt);
            }

            if self.pruning_enabled && scheduled.z1.is_empty() {
                debug!("  [剪枝] Z1为空,跳过此状态");
                self.stats.pruned_states_count += 1;
                continue;
            } else if !self.pruning_enabled && scheduled.z1.is_empty() {
                debug!("  [警告] Z1为空,但剪枝已禁用,继续处理");
            }

            let mut fired_count = 0;
            for &t in &chosen {
                if let Some((nxt, tau)) = self.fire_with_dbm(t, &scheduled) {
                    debug!("  T{} -> 后继状态: ID={}", t, nxt.state_id);

                    let key = StateKey::from(&nxt);
                    let v = if let Some(&existing) = self.state_to_vertex.get(&key) {
                        existing
                    } else {
                        let canonical_nxt = self.canonicalize(&nxt);
                        let new_v = self.find_or_add_vertex(&nxt);
                        queue.push_back(canonical_nxt);
                        self.state_to_vertex.insert(key, new_v);
                        self.stats.total_states += 1;
                        debug!("  [新状态] 添加到图和队列");
                        new_v
                    };

                    self.graph.add_edge(u, v, TransitionEdge::new(t as i32, tau));
                    self.stats.total_transitions += 1;
                    fired_count += 1;
                } else {
                    if self.pruning_enabled {
                        self.stats.pruned_states_count += 1;
                    }
                }
            }

            info!(
                "[STATE_CLASS] 状态 {}: 选择了 {} 个候选变迁, 成功触发 {} 个, 队列大小: {}, 总状态数: {}",
                cur.state_id,
                chosen.len(),
                fired_count,
                queue.len(),
                self.stats.total_states
            );

            if self.stats.total_states >= max_states {
                break;
            }
        }

        info!(
            "[STATE_CLASS] 构建完成: 总迭代次数={}, 总状态数={}",
            iteration, self.stats.total_states
        );

        self.stats.total_states
    }

    pub fn get_graph(&self) -> &StableDiGraph<StateClass, TransitionEdge> {
        &self.graph
    }

    pub fn get_initial_vertex(&self) -> Option<petgraph::graph::NodeIndex> {
        self.initial_vertex
    }

    pub fn get_statistics(&self) -> &ReachabilityStats {
        &self.stats
    }

    fn create_initial_state_class(&mut self) -> StateClass {
        let mut initial = StateClass::new();
        initial.marking = self.ptpn.get_marking().clone();
        initial.state_id = self.next_state_id;
        self.next_state_id += 1;
        initial.cumulative_time = 0.0;

        let num_transitions = self.ptpn.num_transitions();
        initial.z1.resize(num_transitions + 1);
        initial.z2.resize(num_transitions + 1);

        self.compute_enabled_and_clocks(&mut initial);

        debug!("[STATE_CLASS] 创建初始状态");
        debug!("  标识: {:?}", initial.marking);
        debug!("  使能变迁: {:?}", initial.enabled);

        initial
    }

    fn select_per_core(&self, enabled: &HashSet<usize>) -> Vec<usize> {
        let mut best_per_core: HashMap<i32, usize> = HashMap::new();

        for &t in enabled {
            if let Some(trans) = self.ptpn.get_transition(t) {
                let core = trans.core;
                let priority = trans.priority;
                let entry = best_per_core.entry(core).or_insert(t);
                if let Some(other) = self.ptpn.get_transition(*entry) {
                    if priority > other.priority {
                        *entry = t;
                    }
                }
            }
        }

        let mut chosen: Vec<usize> = best_per_core.values().copied().collect();
        chosen.sort();
        chosen
    }

    fn apply_preemption(&self, chosen: &[usize], state: &mut StateClass) {
        state.suspended.clear();

        let num_transitions = self.ptpn.num_transitions();
        for u in 0..num_transitions {
            let Some(trans_u) = self.ptpn.get_transition(u) else {
                continue;
            };
            if !trans_u.suspendable {
                continue;
            }

            for &t in chosen {
                let Some(trans_t) = self.ptpn.get_transition(t) else {
                    continue;
                };
                if trans_t.core == trans_u.core && trans_t.priority > trans_u.priority {
                    state.suspended.insert(u);
                    let clock_idx = u + 1;
                    if clock_idx < state.z1.size() && clock_idx < state.z2.size() {
                        state.z1.copy_clock_constraints(clock_idx, &mut state.z2);
                        state.z1.freeze_clock(clock_idx);
                        state.z2.freeze_clock(clock_idx);
                    }
                    debug!("    T{}: 被 T{} 抢占,冻结时钟", u, t);
                    break;
                }
            }
        }
    }

    fn maximal_time_elapse(&self, state: &mut StateClass, dt: &mut f64) -> bool {
        let mut ub_star = INF_TIME;

        let num_transitions = self.ptpn.num_transitions();

        for i in 1..state.z1.size().min(num_transitions + 1) {
            if state.z1.is_frozen(i) {
                continue;
            }
            let ub = state.z1.get_constraint(i, 0);
            if ub != INF_TIME && ub < ub_star {
                ub_star = ub;
            }
        }

        for i in 1..state.z2.size().min(num_transitions + 1) {
            if state.z2.is_frozen(i) {
                continue;
            }
            let ub = state.z2.get_constraint(i, 0);
            if ub != INF_TIME && ub < ub_star {
                ub_star = ub;
            }
        }

        if ub_star == INF_TIME || ub_star <= 0 {
            *dt = 0.0;
            return false;
        }

        state.z1.elapse_time(ub_star);
        state.z2.elapse_time(ub_star);
        state.cumulative_time += ub_star as f64;
        *dt = ub_star as f64;

        debug!("  最大化时间推进: dt = {}", dt);
        true
    }

    fn fire_with_dbm(
        &mut self,
        trans_idx: usize,
        from_state: &StateClass,
    ) -> Option<(StateClass, f64)> {
        let mut to = from_state.copy();
        to.state_id = self.next_state_id;
        self.next_state_id += 1;

        let trans = self.ptpn.get_transition(trans_idx)?;
        let alpha = trans.time_interval.earliest;
        let beta = if trans.time_interval.latest == INF {
            INF_TIME
        } else {
            trans.time_interval.latest
        };

        let target_z = if trans.suspendable { &to.z2 } else { &to.z1 };
        let zcheck = target_z.restrict_for_firing(trans_idx, alpha, beta);

        if self.pruning_enabled {
            if zcheck.is_empty() || !zcheck.is_consistent() {
                debug!("    T{}: 触发窗口检查失败", trans_idx);
                return None;
            }
        } else if zcheck.is_empty() || !zcheck.is_consistent() {
            debug!("    T{}: 触发窗口检查失败,但剪枝已禁用", trans_idx);
        }

        let delta = if alpha == 0 && beta == 0 {
            0
        } else if alpha == beta {
            alpha
        } else {
            alpha
        };

        if delta > 0 {
            to.z1.elapse_time(delta);
            to.z2.elapse_time(delta);
            to.cumulative_time += delta as f64;
        }

        let fire_time = to.cumulative_time;

        to.marking = MatrixPTPN::fire(&to.marking, &self.ptpn, trans_idx).ok()?;
        if to.marking.is_empty() {
            debug!("    T{}: 触发后标识为空", trans_idx);
            return None;
        }

        let clock_idx = trans_idx + 1;
        if trans.suspendable {
            if clock_idx < to.z2.size() {
                to.z2.reset_clock(clock_idx);
            }
            to.suspended.remove(&trans_idx);
        } else {
            if clock_idx < to.z1.size() {
                to.z1.reset_clock(clock_idx);
            }
        }

        self.compute_enabled_and_clocks(&mut to);

        debug!("    T{}: 成功触发, fire_time = {}", trans_idx, fire_time);
        Some((to, fire_time))
    }

    fn compute_enabled_and_clocks(&self, state: &mut StateClass) {
        let num_transitions = self.ptpn.num_transitions();

        let mut new_enabled = HashSet::new();
        for t in 0..num_transitions {
            if MatrixPTPN::is_enabled(&state.marking, &self.ptpn, t) {
                new_enabled.insert(t);
            }
        }

        state.z1.resize(num_transitions + 1);
        state.z2.resize(num_transitions + 1);

        let old_enabled = state.enabled.clone();
        for t in &old_enabled {
            if !new_enabled.contains(t) {
                let clock_idx = t + 1;
                if clock_idx < state.z1.size() {
                    state.z1.reset_clock(clock_idx);
                }
                if clock_idx < state.z2.size() {
                    state.z2.reset_clock(clock_idx);
                }
                state.z1.unfreeze_clock(clock_idx);
                state.z2.unfreeze_clock(clock_idx);
            }
        }

        state.enabled = new_enabled.clone();

        for &trans_idx in &new_enabled {
            let trans = match self.ptpn.get_transition(trans_idx) {
                Some(t) => t,
                None => continue,
            };
            let clock_idx = trans_idx + 1;

            let alpha = trans.time_interval.earliest;
            let beta = if trans.time_interval.latest == INF {
                INF_TIME
            } else {
                trans.time_interval.latest
            };

            if trans.suspendable {
                if alpha > 0 {
                    state.z2.set_constraint(0, clock_idx, -alpha);
                } else {
                    state.z2.set_constraint(0, clock_idx, 0);
                }
                if beta != INF_TIME {
                    state.z2.set_constraint(clock_idx, 0, beta);
                } else {
                    state.z2.set_constraint(clock_idx, 0, INF_TIME);
                }
            } else {
                if alpha > 0 {
                    state.z1.set_constraint(0, clock_idx, -alpha);
                } else {
                    state.z1.set_constraint(0, clock_idx, 0);
                }
                if beta != INF_TIME {
                    state.z1.set_constraint(clock_idx, 0, beta);
                } else {
                    state.z1.set_constraint(clock_idx, 0, INF_TIME);
                }
            }
        }

        self.recompute_suspension(state);
    }

    fn recompute_suspension(&self, state: &mut StateClass) {
        let enabled_vec: Vec<usize> = state.enabled.iter().copied().collect();
        let mut suspended = HashSet::new();

        for &t in &state.enabled {
            if self.is_suspended(t, &enabled_vec) {
                suspended.insert(t);
                if !state.suspended.contains(&t) {
                    let clock_idx = t + 1;
                    state.z1.copy_clock_constraints(clock_idx, &mut state.z2);
                    state.z1.freeze_clock(clock_idx);
                    state.z2.freeze_clock(clock_idx);
                }
            } else if state.suspended.contains(&t) {
                let clock_idx = t + 1;
                state.z2.copy_clock_constraints(clock_idx, &mut state.z1);
                state.z1.unfreeze_clock(clock_idx);
                state.z2.unfreeze_clock(clock_idx);
            }
        }

        state.suspended = suspended;
    }

    fn is_suspended(&self, trans_idx: usize, enabled: &[usize]) -> bool {
        let trans = match self.ptpn.get_transition(trans_idx) {
            Some(t) => t,
            None => return false,
        };
        if !trans.suspendable {
            return false;
        }

        let trans_core = trans.core;
        let trans_priority = trans.priority;

        for &other_t in enabled {
            if other_t == trans_idx {
                continue;
            }
            if let Some(other) = self.ptpn.get_transition(other_t) {
                if other.core == trans_core
                    && !other.suspendable
                    && other.priority > trans_priority
                {
                    return true;
                }
            }
        }
        false
    }

    fn canonicalize(&self, state: &StateClass) -> StateClass {
        let mut canonical = state.copy();
        canonical.z1.minimize();
        canonical.z2.minimize();
        canonical
    }

    fn find_or_add_vertex(&mut self, state: &StateClass) -> petgraph::graph::NodeIndex {
        let key = StateKey::from(state);
        if let Some(&v) = self.state_to_vertex.get(&key) {
            return v;
        }
        let v = self.graph.add_node(state.copy());
        self.state_to_vertex.insert(key, v);
        v
    }

    pub fn save_to_dot(&self, path: impl AsRef<Path>) -> std::io::Result<()> {
        let mut dot = String::from("digraph StateClassGraph {\n");

        for (idx, node) in self.graph.node_weights().enumerate() {
            let label = format!(
                "S{} [M={:?}]",
                node.state_id,
                node.marking
            );
            dot.push_str(&format!("  n{} [label=\"{}\"];\n", idx, label.replace('"', "'")));
        }

        for edge in self.graph.edge_references() {
            let from = edge.source().index();
            let to = edge.target().index();
            let w = edge.weight();
            dot.push_str(&format!(
                "  n{} -> n{} [label=\"T{}@{:.1}\"];\n",
                from, to, w.transition_id, w.firing_time
            ));
        }

        dot.push_str("}\n");
        fs::write(path, dot)
    }

    pub fn save_to_json(&self, path: impl AsRef<Path>) -> std::io::Result<()> {
        let mut nodes = Vec::new();
        for (idx, node) in self.graph.node_weights().enumerate() {
            nodes.push(serde_json::json!({
                "id": idx,
                "state_id": node.state_id,
                "marking": node.marking,
                "cumulative_time": node.cumulative_time,
            }));
        }

        let mut edges = Vec::new();
        for edge in self.graph.edge_references() {
            edges.push(serde_json::json!({
                "from": edge.source().index(),
                "to": edge.target().index(),
                "transition_id": edge.weight().transition_id,
                "firing_time": edge.weight().firing_time,
            }));
        }

        let output = serde_json::json!({
            "nodes": nodes,
            "edges": edges,
            "stats": {
                "total_states": self.stats.total_states,
                "total_transitions": self.stats.total_transitions,
            }
        });

        fs::write(path, serde_json::to_string_pretty(&output)?)
    }
}

File: src/analysis/scheduling/reachability.cpp
#include "reachability.h"

#include <algorithm>
#include <fstream>
#include <iomanip>
#include <limits>

namespace scheduling {

ReachabilityGraph::ReachabilityGraph(const petri::PTPN& ptpn)
    : ptpn_(ptpn), next_state_id_(0) {}

StateKey ReachabilityGraph::make_key(const StateClass& state) const {
  StateKey key;
  key.marking = state.marking;
  key.zone_matrix = state.zone.raw_matrix();
  key.frozen_clocks = state.zone.frozen_clocks();
  key.enabled = state.enabled;
  key.active = state.active;
  key.suspended = state.suspended;
  return key;
}

ReachabilityGraph::Vertex ReachabilityGraph::find_or_add_state(const StateClass& state) {
  StateKey key = make_key(state);
  auto it = state_to_vertex_.find(key);
  if (it != state_to_vertex_.end()) {
    stats_.dedup_hits++;
    return it->second;
  }

  stats_.dedup_misses++;
  StateClass new_state = state;
  new_state.state_id = next_state_id_++;

  Vertex v = states_.size();
  states_.push_back(new_state);
  state_to_vertex_.emplace(std::move(key), v);

  return v;
}

std::set<size_t> ReachabilityGraph::compute_enabled(
    const std::vector<int>& marking) const {
  std::set<size_t> enabled;
  const size_t num_transitions = ptpn_.num_transitions();

  for (size_t t = 0; t < num_transitions; ++t) {
    if (petri::PTPN::is_enabled(marking, ptpn_, t)) {
      enabled.insert(t);
    }
  }

  return enabled;
}

int ReachabilityGraph::compute_firing_time(
    const StateClass& state, size_t t) const {
  if (!state.enabled.count(t)) {
    return -1;
  }

  if (state.suspended.count(t)) {
    return -1;  // 挂起的变迁不能发生
  }

  if (!state.has_clock_for_transition(t)) {
    return -1;
  }

  const auto& trans = ptpn_.get_transition(t);
  const int alpha = trans.time_interval.earliest;
  const int beta = trans.time_interval.latest == petri::INF
                       ? INF_TIME
                       : trans.time_interval.latest;

  const size_t clock_idx = static_cast<size_t>(
      state.clock_index_for_transition(t));
  const int lower_bound = state.zone.get_lower_bound(clock_idx);
  const int upper_bound = state.zone.get_upper_bound(clock_idx);

  // τ(t) = max(α(t), lb_t)
  const int firing_time = std::max(alpha, lower_bound);

  // 检查是否在时间窗口内
  if (firing_time > upper_bound) {
    return -1;
  }
  if (beta != INF_TIME && firing_time > beta) {
    return -1;
  }

  return firing_time;
}

void ReachabilityGraph::advance_time(StateClass& state, int delta) const {
  if (delta <= 0) return;

  // 推进所有非冻结的 active 时钟
  state.zone.elapse_time(delta);
  state.cumulative_time += delta;
}

size_t ReachabilityGraph::build(size_t max_states) {
  stats_ = Statistics();
  states_.clear();
  edges_.clear();
  state_to_vertex_.clear();
  next_state_id_ = 0;

  // 创建初始状态
  StateClass initial;
  initial.marking = ptpn_.get_marking();
  initial.cumulative_time = 0.0;
  initial.zone = DBM(1);  // 只有参考时钟 c₀

  recompute_sets(initial);

  Vertex initial_vertex = find_or_add_state(initial);
  stats_.total_states++;

  std::vector<Vertex> frontier{initial_vertex};

  // BFS 展开
  while (!frontier.empty()) {
    if (stats_.total_states >= max_states) {
      stats_.truncated = true;
      break;
    }

    std::vector<Vertex> next_frontier;

    for (Vertex v : frontier) {
      const StateClass& cur = states_[v];
      std::vector<StateClass> successors = expand(cur);

      for (auto& succ : successors) {
        const StateKey succ_key = make_key(succ);
        const bool is_new_state = state_to_vertex_.find(succ_key) == state_to_vertex_.end();
        Vertex succ_vertex = find_or_add_state(succ);
        edges_.push_back({v, succ_vertex});
        stats_.total_edges++;

        if (is_new_state) {
          stats_.total_states++;
          next_frontier.push_back(succ_vertex);
        }
      }
    }

    frontier = std::move(next_frontier);
  }

  return stats_.total_states;
}

const StateClass& ReachabilityGraph::get_state(Vertex v) const {
  return states_.at(v);
}

const StateClass& ReachabilityGraph::get_initial_state() const {
  return states_.front();
}

bool ReachabilityGraph::save_to_dot(const std::string& path) const {
  std::ofstream out(path);
  if (!out.is_open()) return false;

  out << "digraph ReachabilityGraph {\n";
  out << "  rankdir=LR;\n";
  out << "  node [shape=box];\n\n";

  for (size_t i = 0; i < states_.size(); ++i) {
    const auto& s = states_[i];
    out << "  s" << s.state_id << " [label=\"State " << s.state_id << "\\n";
    out << "M: [";
    for (size_t j = 0; j < s.marking.size(); ++j) {
      if (j > 0) out << ",";
      out << s.marking[j];
    }
    out << "]\\n";
    out << "Active: " << s.active.size() << "\\n";
    out << "Time: " << std::fixed << std::setprecision(2) << s.cumulative_time;
    out << "\"];\n";
  }

  out << "\n";

  for (const auto& e : edges_) {
    const auto& src = states_[e.first];
    const auto& tgt = states_[e.second];
    out << "  s" << src.state_id << " -> s" << tgt.state_id << ";\n";
  }

  out << "}\n";
  out.close();

  return true;
}

bool ReachabilityGraph::save_to_json(const std::string& path) const {
  std::ofstream out(path);
  if (!out.is_open()) return false;

  out << "{\n";
  out << "  \"states\": [\n";

  for (size_t i = 0; i < states_.size(); ++i) {
    const auto& s = states_[i];
    if (i > 0) out << ",\n";
    out << "    {\"id\": " << s.state_id << ", \"marking\": [";
    for (size_t j = 0; j < s.marking.size(); ++j) {
      if (j > 0) out << ", ";
      out << s.marking[j];
    }
    out << "], \"active_count\": " << s.active.size() << "}";
  }

  out << "\n  ],\n";
  out << "  \"transitions\": [\n";

  for (size_t i = 0; i < edges_.size(); ++i) {
    const auto& e = edges_[i];
    if (i > 0) out << ",\n";
    out << "    {\"source\": " << states_[e.first].state_id
        << ", \"target\": " << states_[e.second].state_id << "}";
  }

  out << "\n  ],\n";
  out << "  \"statistics\": {"
      << "\"total_states\": " << stats_.total_states
      << ", \"total_edges\": " << stats_.total_edges
      << ", \"dedup_hits\": " << stats_.dedup_hits
      << ", \"dedup_misses\": " << stats_.dedup_misses
      << ", \"truncated\": " << (stats_.truncated ? "true" : "false")
      << "}\n";
  out << "}\n";

  out.close();
  return true;
}

void ReachabilityGraph::recompute_sets(StateClass& state) {
  // 重新计算使能集合
  state.enabled = compute_enabled(state.marking);

  // 计算活跃集合（每核最高优先级）
  state.active = SchedulingAlgorithms::select_active_per_core(
      state.enabled, ptpn_);

  // 计算挂起集合
  state.suspended = SchedulingAlgorithms::compute_suspended(
      state.enabled, state.active, ptpn_);

  // 更新 DBM 时钟映射
  // 添加新使能变迁的时钟
  for (size_t t : state.enabled) {
    if (!state.has_clock_for_transition(t)) {
      // 添加新时钟
      size_t clock_idx = state.zone.add_clock();
      if (state.transition_to_clock.size() <= t) {
        state.transition_to_clock.resize(t + 1, -1);
      }
      state.transition_to_clock[t] = static_cast<int>(clock_idx);
      if (state.clock_to_transition.size() <= clock_idx) {
        state.clock_to_transition.resize(
            clock_idx + 1, std::numeric_limits<size_t>::max());
      }
      state.clock_to_transition[clock_idx] = t;

      // 设置初始约束
      const auto& trans = ptpn_.get_transition(t);
      state.zone.set_constraint(0, clock_idx, -trans.time_interval.earliest);
      if (trans.time_interval.latest != petri::INF) {
        state.zone.set_constraint(clock_idx, 0, trans.time_interval.latest);
      }
    }
  }

  // 冻结/解冻时钟
  for (size_t t : state.enabled) {
    if (state.has_clock_for_transition(t)) {
      size_t clock_idx = static_cast<size_t>(
          state.clock_index_for_transition(t));

      if (state.active.count(t)) {
        state.zone.unfreeze_clock(clock_idx);
      } else {
        state.zone.freeze_clock(clock_idx);
      }
    }
  }

  state.zone.minimize();
}

std::vector<StateClass> ReachabilityGraph::expand(const StateClass& state) {
  std::vector<StateClass> successors;

  // 1. 计算使能集合 E（已在 state 中）
  // 2. 计算活跃集合 X（已在 state 中）
  // 3. 计算挂起集合 R（已在 state 中）

  // 4. 计算最早发生时间
  int tau_min = INF_TIME;
  for (size_t t : state.active) {
    if (state.suspended.count(t)) continue;
    int tau = compute_firing_time(state, t);
    if (tau >= 0) {
      tau_min = std::min(tau_min, tau);
    }
  }

  if (tau_min == INF_TIME) {
    return successors;  // 无可发生变迁
  }

  // 5. 筛选可发生变迁 F = {t | τ(t) = tau_min}
  std::set<size_t> F;
  for (size_t t : state.active) {
    if (state.suspended.count(t)) continue;
    int tau = compute_firing_time(state, t);
    if (tau == tau_min) {
      F.insert(t);
    }
  }

  // 6. 在 F 上取每核最高优先级 X'
  std::set<size_t> X_prime = SchedulingAlgorithms::select_active_per_core(
      F, ptpn_);

  // 7. 从 X' 选一个变迁发生
  if (X_prime.empty()) {
    return successors;
  }

  size_t chosen = SchedulingAlgorithms::select_one_transition(X_prime, ptpn_);

  // 8. 激发变迁，生成后继
  auto successor = fire_transition(state, chosen, tau_min);
  if (successor) {
    successors.push_back(*successor);
  }

  return successors;
}

std::optional<StateClass> ReachabilityGraph::fire_transition(
    const StateClass& state, size_t transition_id, int firing_time) {

  const auto& trans = ptpn_.get_transition(transition_id);

  // 复制状态
  StateClass next = state.copy();

  // 1. 推进时间（仅 active 时钟）
  advance_time(next, firing_time);

  // 2. 更新 DBM：添加时间约束
  if (next.has_clock_for_transition(transition_id)) {
    size_t clock_idx = static_cast<size_t>(
        next.clock_index_for_transition(transition_id));

    // α(t) ≤ c_t ≤ β(t)
    next.zone.set_constraint(0, clock_idx, -trans.time_interval.earliest);
    if (trans.time_interval.latest != petri::INF) {
      next.zone.set_constraint(clock_idx, 0, trans.time_interval.latest);
    }
    next.zone.minimize();
  }

  // 3. 检查 DBM 是否为空
  if (next.zone.is_empty()) {
    return std::nullopt;
  }

  // 4. 激发变迁，更新标识
  next.marking = petri::PTPN::fire(next.marking, ptpn_, transition_id);
  if (next.marking.empty()) {
    return std::nullopt;
  }

  // 5. 重新计算使能/活跃/挂起集合
  recompute_sets(next);

  // 6. 更新累计时间
  next.cumulative_time = state.cumulative_time + firing_time;

  return next;
}

}  // namespace scheduling